aim/web/ui/src/hooks/useDepthMap.tsx

```diff
@@ -6,43 +6,59 @@ import _ from 'lodash-es';
  * @param data {Array<T>} - Array of items which needs for grouping and calculating the initial depth map
  * @param groupItemCb {(item: T) => string} - callback function which calls for every item when grouping the data
  * @param state {any} - engine depthMap state
- * @param deps {unknown[]} - the dependency array, after changing the item of the array will reset the depth map. The default value is: [data]
+ * @param deps {unknown[]} - the dependency array, after changing the item of the array will reset the depth map. The default value is: []
+ * @param sync {boolean} - it enables the depth map synchronization mechanism. The default value is: false
  *
- * return [depthMap, onDepthMapChange] {[Record<string, number>, (value: number, groupKey: string) => void]}
+ * return [depthSelector, onDepthMapChange] {IUseDepthMap}
  */
 
 interface IUseDepthMapProps<T> {
   data: Array<T>;
   groupItemCb: (item: T) => string;
   state: any;
   deps: unknown[];
+  sync?: boolean;
 }
 
+type IUseDepthMap = [
+  (groupId: string) => (state: any) => number,
+  (value: number, groupId: string) => void,
+];
+
 function useDepthMap<T>({
   data,
   groupItemCb,
   state,
-  deps = [data],
-}: IUseDepthMapProps<T>): [
-  (groupKey: string) => (state: any) => number,
-  (value: number, groupKey: string) => void,
-] {
-  const initialMap = React.useMemo(() => {
-    const grouped = _.groupBy(data, groupItemCb);
-    const initial: Record<string, number> = {};
-    Object.keys(grouped).forEach((groupKey: string) => (initial[groupKey] = 0));
-    return initial;
-  }, [data]);
+  deps = [],
+  sync = false,
+}: IUseDepthMapProps<T>): IUseDepthMap {
+  const grouped = _.groupBy(data, groupItemCb);
+
+  const generateMapBy = React.useCallback((value: number = 0) => {
+    const newDepthMap: Record<string, number> = {};
+    for (let [groupId, items] of Object.entries(grouped)) {
+      const maxValue = items.length - 1;
+      newDepthMap[groupId] = value > maxValue ? maxValue : value;
+    }
+    return newDepthMap;
+  }, deps);
```

Comment:
Why dependencies comes from outside?, you know everything `data`, `groupedItemCallback` in your memoized function

Reply:
we don't need to update `generateMapBy` function every time when data changed as it only works when `deps` updates